### PR TITLE
Add port forwarding for rkt with kvm stage1

### PR DESCRIPTION
**Pull request by jjlakis**
Port forwarding for rkt kvm using `socat`.
cc @yifan-gu @euank @pskrzyns @lukasredynk 

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32126)

### CODE REVIEW

**Comment by euank** [comment]
We should also document that there's a slight behavioural difference here in that `port-forward` only works for applications listening on `eth0` within the pod, not `eth0` or `lo` within the pod.

Correct me if I'm wrong ofcourse.

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -2103,20 +2102,28 @@ func (r *Runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
...
 	}
 
-	args := []string{"-t", fmt.Sprintf("%d", listResp.Pods[0].Pid), "-n", socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
-
-	nsenterPath, lookupErr := exec.LookPath("nsenter")
-	if lookupErr != nil {
-		return fmt.Errorf("unable to do port forwarding: nsenter not found.")
+	var args []string
+	var fwCaller string
+	if strings.Contains(r.config.Stage1Image, "kvm") {
```
Comment (euank) [inline pkg/kubelet/rkt/rkt.go]: This won't work in the case that the alpha "stage1 override" annotation is used.

Does the rkt api-service return what stage1 a pod is using? If it doesn't, it probably should.

I'd also slightly prefer checking for a more or less exact match, not just `string.Contains` for this.

File: pkg/kubelet/rkt/rkt.go [hunk 2]
```diff
@@ -2103,20 +2102,28 @@ func (r *Runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
...
+		args = []string{"-", fmt.Sprintf("TCP4:%s:%d", listPod.GetNetworks()[0].Ipv4, port)}
```
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: Probably safer to check `GetNetworks()`'s return value.

File: pkg/kubelet/rkt/rkt.go (hunk 1 above)
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: @euank api service doesn't return the stage1 info today unfortunately...
@jjlakis To workaround this, we can do something like https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/kubelet/rkt/rkt.go#L1043-L1053

Comment (jjlakis) [inline pkg/kubelet/rkt/rkt.go]: @yifan-gu As far as I see, I am unable to get pod manifest in this scope. All pod information I got is [kubecontainer.pod](https://godoc.org/k8s.io/kubernetes/pkg/kubelet/container#Pod), or at most [manifest from api-service](https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/kubelet/rkt/rkt.go#L598). Am I wrong?

Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: @jjlakis `manifest from api-service` is an marshaled pod manifest. We can unmarshal it into a pod manifest.

Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: @jjlakis But we don't need to do that, the pod returned by api service has the annotation. So we can combine the annotation and the `runtime.Config` to check the stage1 image.

Comment (jjlakis) [inline pkg/kubelet/rkt/rkt.go]: @yifan-gu PTAL. @euank: You're right, where we should document this issue?

File: pkg/kubelet/rkt/rkt.go [hunk 3]
```diff
@@ -2122,20 +2121,41 @@ func (r *Runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
...
+	listPod := listResp.Pods[0]
 
 	socatPath, lookupErr := exec.LookPath("socat")
 	if lookupErr != nil {
 		return fmt.Errorf("unable to do port forwarding: socat not found.")
 	}
 
-	args := []string{"-t", fmt.Sprintf("%d", listResp.Pods[0].Pid), "-n", socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
+	// Check in config and in annotations if we're running kvm flavor
+	isKvm := strings.Contains(r.config.Stage1Image, "kvm")
```
Comment (euank) [inline pkg/kubelet/rkt/rkt.go]: This breaks in the reverse now; if you have `stage1-image=kvm` and then have the annotation for non-kvm, `isKvm` still remains true.

Comment (jjlakis) [inline pkg/kubelet/rkt/rkt.go]: @euank I don't see any mistake, there's a loop below that overriding `isKvm` if there's another image in annotations

**Comment by jjlakis** [comment]
PTAL @pskrzyns

**Comment by yifan-gu** [comment]
LGTM, leave @pskrzyns for the final review.

**Comment by pskrzyns** [comment]
LGTM

We are OK with issue that it  will work only for applications listening on ```eth0```.
Full support will be implemented in rktlet.

**Comment by yifan-gu** [comment]
@k8s-bot test this please , issue #IGNORE

**Comment by yifan-gu** [comment]
@k8s-bot gce e2e test this

**Pull request merged**